garden_ai/client.py

```diff
@@ -77,18 +80,41 @@ def __init__(
             NativeAppAuthClient(self.client_id) if not auth_client else auth_client
         )
 
-        self.authorizer = self._authenticate()
-        self.searchauthorizer = self._create_search_authorizer()
+        self.groups_authorizer = self._create_authorizer(GroupsClient.scopes.resource_server)
+        self.search_authorizer = self._create_authorizer(SearchClient.scopes.resource_server)
         self.search_client = (
-            SearchClient(authorizer=self.searchauthorizer)
+            SearchClient(authorizer=self.search_authorizer)
             if not search_client
             else search_client
         )
-        self.garden_authorizer = self._create_garden_authorizer()
+        self.garden_authorizer = self._create_authorizer(GardenClient.scopes.resource_server)
+
+        self._set_up_mlflow_env(self.garden_authorizer.access_token)
+
+    def __del__(self):
+        self._tear_down_mlflow_env()
+
+    def _set_up_mlflow_env(self, garden_access_token: str):
+        self._existing_mlflow_token = os.environ.get('MLFLOW_TRACKING_TOKEN')
+        self._existing_mlflow_uri = os.environ.get('MLFLOW_TRACKING_URI')
+        os.environ['MLFLOW_TRACKING_TOKEN'] = garden_access_token
+        os.environ['MLFLOW_TRACKING_URI'] = GARDEN_ENDPOINT + '/mlflow'
+
+    def _tear_down_mlflow_env(self):
+        if hasattr(self, '_existing_mlflow_token') and self._existing_mlflow_token:
+            os.environ['MLFLOW_TRACKING_TOKEN'] = self._existing_mlflow_token
+        elif os.environ.get('MLFLOW_TRACKING_TOKEN'):
+            del os.environ['MLFLOW_TRACKING_TOKEN']
+        if hasattr(self, '_existing_mlflow_uri') and self._existing_mlflow_uri:
+            os.environ['MLFLOW_TRACKING_URI'] = self._existing_mlflow_uri
+        elif os.environ.get('MLFLOW_TRACKING_URI'):
+            del os.environ['MLFLOW_TRACKING_URI']
```

Comment:
Why all the defensive checks? Apparently you can't reliably mock something called from a Python destructor because the mocks get deregistered before the body of `__del__` is invoked. And I do want to mock out `_set_up_mlflow_env`. So _tear_down_mlflow_env handles the case where those env vars have never been set. It's kludgey and there's probably a better way. But I was rabbit holing on this for a while and want to cut my losses for now.

Reply:
ohhhhh man, for some reason I had assumed it stomped on the variables in the calling process. But that assumption came out of the ether and I never tested it. I'll double check what you're saying but yeah in that case I will gladly go kill this ugly code.

Reply:
lol yep, this is my facepalm of the week. Gonna go kill that code.

```
❯ export TEST=foo
❯ echo $TEST
foo
❯ python
>>> import os
>>> os.environ['TEST'] = 'bar'
>>> ^D
❯ echo $TEST
foo
```